src/compiler/checker.ts

```diff
@@ -16515,7 +16515,8 @@ namespace ts {
                     return mapTypeWithAlias(getReducedType(mappedTypeVariable), t => {
                         if (t.flags & (TypeFlags.AnyOrUnknown | TypeFlags.InstantiableNonPrimitive | TypeFlags.Object | TypeFlags.Intersection) && t !== wildcardType && !isErrorType(t)) {
                             if (!type.declaration.nameType) {
-                                if (isArrayType(t)) {
+                                let constraint;
+                                if (isArrayType(t) || (t.flags & TypeFlags.Any) && (constraint = getConstraintOfTypeParameter(typeVariable)) && everyType(constraint, or(isArrayType, isTupleType))) {

```

Comment:
```suggestion
                                if (isArrayType(t) || (t.flags & TypeFlags.Any) && (constraint = getConstraintOfTypeParameter(typeVariable)) && everyType(constraint, isArrayOrTupleLikeType)) {

```
? Or is including the x-like types (length properties and 0 properties) going to mess up the behavior in a case we care about? I'm thinking passing `any` to a `T extends {length: number}` may also justify an array mapping rather than a `string` index signature.

Reply:
I thought about the tuple-like type thing, and that's one concern. The bigger concern is the implementation of `isArrayLikeType`:

```ts
return isArrayType(type) || !(type.flags & TypeFlags.Nullable) && isTypeAssignableTo(type, anyReadonlyArrayType);
```

which I think would consider `any` an array-like type.

Reply:
Raw `any` in a constraint position is eagerly replaced with `unknown` nowadays (so it behaves as a proper top type constraint and not as an odd anyish thing) so it shouldn't be an issue I don't think.

Reply:
I think that might be fine, but for now there's some precedent in `getResolvedApparentTypeOfMappedType` for the current behavior.